File: src/agent/retriever.js
// 检索适配器：为教学与出题提供**真实依据**。
//
// 依据来源按优先级：
//   ① 用户自己上传的教材（浏览器选文件 → 存本会话内存，不落盘）
//   ② 用户自己的 ima 知识库（凭证随请求传入）
//   ③ 项目内置 corpus/ 目录（开箱即用的示例教材）
import { readdirSync, readFileSync, existsSync, statSync } from 'node:fs'
import { join, resolve, extname } from 'node:path'

const CORPUS_DIR = resolve(process.env.GALGAME_CORPUS || join(process.cwd(), 'corpus'))
const MAX_SECTIONS = 3
const MAX_CHARS = 1200
const IMA_HOST = 'ima.qq.com'
const IMA_BASE = '/openapi/wiki/v1'

/** 把文本切成检索用的 token（英文单词 + 中文单字与双字组合） */
function tokenize(text) {
  const s = String(text || '').toLowerCase()
  const tokens = new Set()
  for (const w of s.match(/[a-z0-9]{2,}/g) || []) tokens.add(w)
  const cjk = s.match(/[\u4e00-\u9fa5]/g) || []
  for (const ch of cjk) tokens.add(ch)
  for (let i = 0; i + 1 < cjk.length; i++) tokens.add(cjk[i] + cjk[i + 1])
  return [...tokens]
}

/** 按 Markdown 标题切成小节 */
function splitSections(text) {
  const lines = String(text || '').split(/\r?\n/)
  const sections = []
  let current = { title: '正文', body: [] }
  for (const line of lines) {
    const m = /^(#{1,6})\s+(.*)$/.exec(line)
    if (m) {
      if (current.body.length) sections.push(current)
      current = { title: m[2].trim(), body: [] }
    } else {
      current.body.push(line)
    }
  }
  if (current.body.length) sections.push(current)
  return sections
}

/** 递归收集语料文件（限 3 层） */
function collectFiles(dir, depth = 0, out = []) {
  if (depth > 3 || !existsSync(dir)) return out
  for (const name of readdirSync(dir)) {
    const full = join(dir, name)
    let st
    try { st = statSync(full) } catch { continue }
    if (st.isDirectory()) collectFiles(full, depth + 1, out)
    else if (['.md', '.txt'].includes(extname(name).toLowerCase())) out.push(full)
  }
  return out
}

/**
 * 在给定文档集合里检索。
 * @param {Array<{name:string, text:string}>} docs 文档集合
 * @param {string} query 查询
 * @param {string} subject 学科（一并参与打分）
 * @param {string} source 来源标签，用于返回值和 UI 提示
 */
function searchDocs(docs, query, subject, source) {
  const tokens = tokenize(`${query || ''} ${subject || ''}`)
  if (tokens.length === 0) return { ok: false, source, count: 0, items: [], error: '查询为空' }
  const scored = []
  for (const doc of docs) {
    for (const sec of splitSections(doc.text)) {
      const body = sec.body.join('\n').trim()
      if (!body) continue
      const hay = `${sec.title}\n${body}`.toLowerCase()
      let score = 0
      for (const t of tokens) if (hay.includes(t)) score += t.length >= 2 ? 2 : 1
      if (score > 0) scored.push({ score, title: sec.title, source: doc.name, content: body.slice(0, MAX_CHARS) })
    }
  }
  scored.sort((a, b) => b.score - a.score)
  const items = scored.slice(0, MAX_SECTIONS).map(({ title, source: src, content }) => ({ title, source: src, content }))
  return { ok: items.length > 0, source, count: items.length, items }
}

/** ① 用户上传的教材（本会话内存）
 *  按学科取用：「通用」（未指定学科）的教材对所有学科可用；
 *  指定了学科的教材**只**对该学科可用。若该学科下无任何可用教材，则如实返回空。 */
export function sessionCorpusRetrieve(query, subject, uploads = []) {
  const all = (uploads || []).filter((f) => f && typeof f.text === 'string' && f.text.trim())
  if (all.length === 0) {
    return { ok: false, source: 'upload', count: 0, items: [], error: '尚未上传教材' }
  }
  let pool = all
  if (subject) {
    pool = all.filter((f) => !f.subject || f.subject === subject)
    if (pool.length === 0) {
      return {
        ok: false, source: 'upload', count: 0, items: [],
        error: `该学科（${subject}）没有可用教材：你上传的文件都指定了别的学科，且没有「通用」教材`,
      }
    }
  }
  const docs = pool.map((f) => ({
    name: f.subject ? `${f.name}（${f.subject}）` : (f.name || '上传教材'),
    text: f.text,
  }))
  return searchDocs(docs, query, subject, 'upload')
}

/** ③ 项目内置 corpus/ 目录 */
export function localCorpusRetrieve(query, subject) {
  const files = collectFiles(CORPUS_DIR)
  if (files.length === 0) {
    return { ok: false, source: 'corpus', count: 0, items: [], error: `语料目录为空：${CORPUS_DIR}（可放入 .md/.txt 教材文件）` }
  }
  const docs = []
  for (const file of files) {
    try { docs.push({ name: file, text: readFileSync(file, 'utf8') }) } catch { /* 跳过读不了的文件 */ }
  }
  return searchDocs(docs, query, subject, 'corpus')
}

/* ══════════ ima 知识库 ══════════ */

/** 解析 ima 凭证（请求优先，环境变量兜底） */
function imaCreds(creds = {}) {
  return {
    apiKey: creds.imaApiKey || process.env.IMA_API_KEY || '',
    clientId: creds.imaClientId || process.env.IMA_CLIENT_ID || '',
  }
}

/** 解析「学科 → 知识库ID」映射（字符串或对象都支持） */
export function parseKbMap(raw) {
  try { return typeof raw === 'string' ? JSON.parse(raw || '{}') : (raw || {}) } catch { return {} }
}

/** ima 开放接口 POST */
async function imaPost(path, body, creds = {}) {
  const { apiKey, clientId } = imaCreds(creds)
  const res = await fetch(`https://${IMA_HOST}${IMA_BASE}${path}`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
      'ima-openapi-clientid': clientId,
      'ima-openapi-apikey': apiKey,
    },
    body: JSON.stringify(body || {}),
  })
  return await res.json()
}

/** ② ima 知识库内检索；未配置或没有可用知识库时返回 null */
/**
 * 判断下载回来的东西是不是「书」的原始二进制（PDF / EPUB / zip / 图片…）。
 *
 * ima 的 get_media_info 给的是**原始文件**地址。往知识库里放的是 PDF/EPUB 时，
 * 拿到的就是几十 MB 二进制；当文本读只会得到 `%PDF-1.6 %äüöß...` 这种乱码。
 * 把乱码喂给模型等于让它读天书 —— 宁可如实说「读不了」，也不能假装读到了。
 */
function binaryKind(buf) {
  const b = buf.subarray(0, 8)
  const hex = Buffer.from(b).toString('latin1')
  if (hex.startsWith('%PDF')) return 'PDF'
  if (hex.startsWith('PK')) return '压缩包（EPUB/Word/Excel）'
  if (hex.startsWith('\x89PNG')) return '图片'
  if (hex.startsWith('\xff\xd8\xff')) return '图片'
  if (hex.startsWith('GIF8')) return '图片'
  if (hex.startsWith('RIFF')) return '音视频'
  if (hex.startsWith('ID3')) return '音频'
  // 出现 NUL 字节基本可以断定不是文本
  const probe = buf.subarray(0, 1024)
  for (let i = 0; i < probe.length; i++) if (probe[i] === 0) return '二进制文件'
  return ''
}

async function imaRetrieve(query, subject, creds = {}) {
  const { apiKey, clientId } = imaCreds(creds)
  const map = parseKbMap(creds.imaKbMap || process.env.IMA_KB_MAP || '')
  const kb = map[subject] || Object.values(map)[0] || creds.imaKb || process.env.IMA_KB || ''
  if (!apiKey || !clientId || !kb) return null

  const sj = await imaPost('/search_knowledge', { knowledge_base_id: kb, query: String(query || ''), cursor: '' }, creds)
  if (sj.code !== 0) return { ok: false, apiOk: false, source: 'ima', items: [], error: sj.msg }
  const hits = (sj.data && sj.data.info_list) || []
  const items = []
  let unreadable = 0
  for (const h of hits.slice(0, 2)) {
    const entry = { title: h.title, source: 'ima', content: '' }
    try {
      const mj = await imaPost('/get_media_info', { media_id: h.media_id }, creds)
      const url = mj && mj.data && mj.data.url_info && mj.data.url_info.url
      if (url) {
        const buf = Buffer.from(await (await fetch(url)).arrayBuffer())
        const kind = binaryKind(buf)
        if (kind) {
          // 是原始文件，不是文本 —— 明确说清楚，不把乱码当正文
          unreadable++
          entry.error = `这是一本 ${kind} 原始文件，需要先抽出正文才能读`
        } else {
          entry.content = buf.toString('utf8').slice(0, MAX_CHARS)
        }
      }
    } catch (e) {
      entry.error = String((e && e.message) || e)
    }
    items.push(entry)
  }
  const usable = items.filter((x) => x.content)
  if (!usable.length && unreadable) {
    return {
      ok: false, apiOk: true, source: 'ima', count: hits.length, items: [],
      error: `知识库里的《${items[0].title || '这本书'}》是 PDF/EPUB 原始文件，Node 版还没做正文抽取`,
      hint: 'ima 自己的搜索只匹配书名、拿不到书里内容。需要「下载原始文件 → 抽取正文 → 本地检索」'
        + '这一整套（Python 版已实现，见 python/agent/textract.py 与 ima_index.py）。'
        + '要么改用 Python 版 / Windows 桌面版，要么把书转成 .md / .txt 再放进知识库。',
    }
  }
  return { ok: usable.length > 0, apiOk: true, source: 'ima', count: hits.length, items: usable }
}

/**
 * 统一检索入口（按优先级回落）。
 * @param {string} query 查询词
 * @param {string} [subject] 学科
 * @param {object} [creds] 本次请求携带的凭证
 * @param {Array<{name:string,text:string}>} [uploads] 本会话上传的教材
 */
export async function retrieve(query, subject, creds = {}, uploads = []) {
  // ① 用户上传的教材优先
  if (uploads && uploads.length > 0) {
    const r = sessionCorpusRetrieve(query, subject, uploads)
    if (r.ok) return r
  }
  // ② ima 知识库
  try {
    const viaIma = await imaRetrieve(query, subject, creds)
    if (viaIma && viaIma.ok !== false) return viaIma
    if (viaIma && viaIma.ok === false && viaIma.apiOk === false) return viaIma // 接口报错，如实返回
  } catch { /* 回落 */ }
  // ③ 内置语料
  return localCorpusRetrieve(query, subject)
}

/** 该次请求是否会走 ima（供 UI 提示用） */
export function imaEnabled(creds = {}) {
  const { apiKey, clientId } = imaCreds(creds)
  return Boolean(apiKey && clientId)
}

/** 列出该凭证可用的 ima 知识库（名称 → ID），供设置面板「按名称选择」 */
export async function listKnowledgeBases(creds = {}) {
  const { apiKey, clientId } = imaCreds(creds)
  if (!apiKey || !clientId) return { ok: false, error: '未填写 ima API Key 或 Client ID' }
  try {
    // 主接口：我可添加的知识库列表（limit ≤ 50）
    // 兜底：知识库搜索（注意 limit 必须在 (0,20]，否则返回 code 51）
    let j = await imaPost('/get_addable_knowledge_base_list', { limit: 50 }, creds)
    if (j.code !== 0) j = await imaPost('/search_knowledge_base', { limit: 20 }, creds)
    if (j.code !== 0) return { ok: false, error: j.msg || `ima 返回 code ${j.code}` }

    const d = j.data || {}
    const raw = d.addable_knowledge_base_list || d.info_list || d.list ||
      d.knowledge_base_list || d.items || (Array.isArray(d) ? d : [])
    const items = []
    for (const it of raw) {
      const id = it.knowledge_base_id || it.kb_id || it.id || it.base_id || ''
      const name = it.name || it.title || it.knowledge_base_name || '(未命名)'
      if (id) items.push({ id, name })
    }
    return { ok: items.length > 0, items, raw: items.length ? undefined : JSON.stringify(j).slice(0, 500) }
  } catch (e) {
    return { ok: false, error: String((e && e.message) || e) }
  }
}

/** 直接用给定凭证测一次 ima 检索（给 UI 的「测试连接」按钮用；不回落到本地语料）
 *  判定标准：接口能正常应答即算连通（命中 0 条只是该测试词没匹配到，不算失败）。 */
export async function testIma(creds = {}) {
  if (!imaEnabled(creds)) {
    return { ok: false, error: '未填写 ima API Key 或 Client ID' }
  }
  try {
    const r = await imaRetrieve('纳什均衡', undefined, creds)
    if (!r) return { ok: false, error: '未指定知识库：请先「拉取知识库列表」并给学科选择知识库' }
    if (r.apiOk === false) return { ok: false, error: r.error || 'ima 接口返回错误' }
    return {
      ...r,
      ok: true,
      count: r.count || 0,
      note: (r.count || 0) > 0
        ? `连接正常，命中 ${r.count} 条`
        : '连接正常（该测试词暂无命中，可换关键词再试）',
    }
  } catch (e) {
    return { ok: false, error: String((e && e.message) || e) }
  }
}

File: python/agent/retriever.py
# -*- coding: utf-8 -*-
"""检索适配器：为教学与出题提供**真实依据**。

依据来源按优先级：
  ① 用户自己上传的教材（浏览器选文件 -> 存本会话内存，不落盘）
  ② 用户自己的 ima 知识库（凭证随请求传入）
  ③ 项目内置 corpus/ 目录（开箱即用的示例教材）
"""

import json
import os
import re
import urllib.error
import urllib.parse
import urllib.request

from . import ima_index

MAX_SECTIONS = 3
MAX_CHARS = 1200
IMA_HOST = 'ima.qq.com'
IMA_BASE = '/openapi/wiki/v1'

CORPUS_DIR = os.path.abspath(
    os.environ.get('GALGAME_CORPUS')
    or os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'corpus'))


def tokenize(text):
    """把文本切成检索用的 token（英文单词 + 中文单字与双字组合）。"""
    s = str(text or '').lower()
    tokens = set()
    for w in re.findall(r'[a-z0-9]{2,}', s):
        tokens.add(w)
    cjk = re.findall(r'[\u4e00-\u9fa5]', s)
    for ch in cjk:
        tokens.add(ch)
    for i in range(len(cjk) - 1):
        tokens.add(cjk[i] + cjk[i + 1])
    return list(tokens)


def split_sections(text):
    """按 Markdown 标题切成小节。"""
    lines = str(text or '').split('\n')
    sections = []
    current = {'title': '正文', 'body': []}
    for line in lines:
        m = re.match(r'^(#{1,6})\s+(.*)$', line)
        if m:
            if current['body']:
                sections.append(current)
            current = {'title': m.group(2).strip(), 'body': []}
        else:
            current['body'].append(line)
    if current['body']:
        sections.append(current)
    return sections


def collect_files(directory, depth=0, out=None):
    """递归收集语料文件（限 3 层）。"""
    out = out if out is not None else []
    if depth > 3 or not os.path.exists(directory):
        return out
    for name in os.listdir(directory):
        full = os.path.join(directory, name)
        try:
            is_dir = os.path.isdir(full)
        except OSError:
            continue
        if is_dir:
            collect_files(full, depth + 1, out)
        elif os.path.splitext(name)[1].lower() in ('.md', '.txt'):
            out.append(full)
    return out


def search_docs(docs, query, subject, source):
    """在给定文档集合里检索。docs: [{'name':..,'text':..}]"""
    tokens = tokenize('%s %s' % (query or '', subject or ''))
    if not tokens:
        return {'ok': False, 'source': source, 'count': 0, 'items': [], 'error': '查询为空'}
    scored = []
    for doc in docs:
        for sec in split_sections(doc['text']):
            body = '\n'.join(sec['body']).strip()
            if not body:
                continue
            hay = ('%s\n%s' % (sec['title'], body)).lower()
            score = 0
            for t in tokens:
                if t in hay:
                    score += 2 if len(t) >= 2 else 1
            if score > 0:
                scored.append({'score': score, 'title': sec['title'],
                               'source': doc['name'], 'content': body[:MAX_CHARS]})
    scored.sort(key=lambda x: -x['score'])
    items = [{'title': x['title'], 'source': x['source'], 'content': x['content']}
             for x in scored[:MAX_SECTIONS]]
    return {'ok': len(items) > 0, 'source': source, 'count': len(items), 'items': items}


def session_corpus_retrieve(query, subject, uploads=None):
    """① 用户上传的教材（本会话内存）。

    按学科取用：「通用」（未指定学科）的教材对所有学科可用；
    指定了学科的教材**只**对该学科可用。若该学科下无任何可用教材，则如实返回空。
    """
    all_docs = [f for f in (uploads or [])
                if f and isinstance(f.get('text'), str) and f['text'].strip()]
    if not all_docs:
        return {'ok': False, 'source': 'upload', 'count': 0, 'items': [], 'error': '尚未上传教材'}
    pool = all_docs
    if subject:
        pool = [f for f in all_docs if not f.get('subject') or f.get('subject') == subject]
        if not pool:
            return {
                'ok': False, 'source': 'upload', 'count': 0, 'items': [],
                'error': '该学科（%s）没有可用教材：你上传的文件都指定了别的学科，且没有「通用」教材' % subject,
            }
    docs = [{'name': ('%s（%s）' % (f['name'], f['subject'])) if f.get('subject')
             else (f.get('name') or '上传教材'),
             'text': f['text']} for f in pool]
    return search_docs(docs, query, subject, 'upload')


def local_corpus_retrieve(query, subject):
    """③ 项目内置 corpus/ 目录。"""
    files = collect_files(CORPUS_DIR)
    if not files:
        return {'ok': False, 'source': 'corpus', 'count': 0, 'items': [],
                'error': '语料目录为空：%s（可放入 .md/.txt 教材文件）' % CORPUS_DIR}
    docs = []
    for path in files:
        try:
            with open(path, 'r', encoding='utf-8', errors='replace') as fh:
                docs.append({'name': path, 'text': fh.read()})
        except OSError:
            continue
    return search_docs(docs, query, subject, 'corpus')


# ══════════ ima 知识库 ══════════

def ima_creds(creds=None):
    """解析 ima 凭证（请求优先，环境变量兜底）。"""
    creds = creds or {}
    return {
        'apiKey': creds.get('imaApiKey') or os.environ.get('IMA_API_KEY') or '',
        'clientId': creds.get('imaClientId') or os.environ.get('IMA_CLIENT_ID') or '',
    }


def parse_kb_map(raw):
    """解析「学科 -> 知识库ID」映射（字符串或对象都支持）。"""
    try:
        return json.loads(raw or '{}') if isinstance(raw, str) else (raw or {})
    except Exception:
        return {}


def _ima_post(path, body, creds=None):
    """ima 开放接口 POST（带限流退避重试）。"""
    c = ima_creds(creds)
    return ima_index._post(path, body, c['apiKey'], c['clientId'])


def ima_retrieve(query, subject, creds=None):
    """② 在 ima 知识库里检索**正文**。

    做法：先用索引器把库里还没下载的书下下来、抽出正文存到本地，
    再在本地做真正的全文检索 —— 因为 ima 自己的 search_knowledge 只匹配标题，
    书里的内容它搜不到，而且它给的是 PDF/EPUB 原始文件，当文本读是乱码。

    未配置 / 没绑定知识库时返回 None（交给上层回落到内置语料）。
    """
    c = ima_creds(creds)
    creds = creds or {}
    kb_map = parse_kb_map(creds.get('imaKbMap') or os.environ.get('IMA_KB_MAP') or '')
    kb = (kb_map.get(subject) or (list(kb_map.values())[0] if kb_map else '')
          or creds.get('imaKb') or os.environ.get('IMA_KB') or '')
    if not c['apiKey'] or not c['clientId'] or not kb:
        return None

    focus = '%s %s' % (query or '', subject or '')
    try:
        stats = ima_index.build_index(kb, c['apiKey'], c['clientId'], focus=focus)
    except ima_index.ImaError as e:
        return {'ok': False, 'apiOk': False, 'source': 'ima', 'items': [],
                'error': 'ima 接口报错：%s' % e}
    except Exception as e:
        return {'ok': False, 'apiOk': False, 'source': 'ima', 'items': [],
                'error': 'ima 索引失败：%s' % e}

    r = ima_index.search_index(kb, query, subject)
    base = {'source': 'ima', 'indexed': r.get('indexed'), 'total': r.get('total'),
            'kbId': kb, 'indexStats': stats}

    if r.get('ok'):
        base.update({'ok': True, 'apiOk': True, 'count': len(r['items']), 'items': r['items'],
                     'via': '本地全文检索（已下载并解析你的书）'})
        return base

    # 没命中 —— 如实说明卡在哪一步，便于用户自己判断
    if r.get('indexed', 0) == 0 and r.get('total', 0) == 0:
        why = '这个知识库里没列出任何文件'
    elif r.get('indexed', 0) == 0:
        why = ('知识库里有 %d 个文件，但一个都没能解析出正文（%s）'
               % (r.get('total', 0), '；'.join(stats.get('notes') or []) or '见索引状态'))
    else:
        why = '已在 %d 本书里全文检索，没有和「%s」相关的内容' % (r.get('indexed'), query or subject or '')
    base.update({'ok': False, 'apiOk': True, 'count': 0, 'items': [], 'error': why,
                 'hint': '知识库状态：已解析 %s 本 / 共 %s 个文件'
                         % (r.get('indexed'), r.get('total'))})
    return base


def retrieve(query, subject=None, creds=None, uploads=None):
    """统一检索入口（按优先级回落）。

    ⚠️ 一个刻意的行为：**配了 ima 时，绝不静默换成项目内置的示例语料**。
    以前会默默回落，于是用户以为老师在讲自己的书，其实讲的是仓库里的两篇示例 ——
    这正是「抓不到我放在知识库里的书」这个感受的来源之一。
    现在回落时会带上 note 说明「这不是你的资料」，让模型如实告诉用户。
    """
    # ① 用户上传的教材优先
    if uploads:
        r = session_corpus_retrieve(query, subject, uploads)
        if r['ok']:
            return r

    # ② ima 知识库（配了才走，且不再无声回落）
    if ima_enabled(creds):
        try:
            via_ima = ima_retrieve(query, subject, creds)
        except Exception as e:
            via_ima = {'ok': False, 'apiOk': False, 'source': 'ima', 'items': [],
                       'error': str(e)}
        if via_ima is not None:
            if via_ima.get('ok'):
                return via_ima
            local = local_corpus_retrieve(query, subject)
            if local.get('ok'):
                local['fallbackFrom'] = 'ima'
                local['note'] = (
                    '⚠️ 以下内容**不是**你的资料，而是项目内置的示例语料。'
                    '你的 ima 知识库没给出结果，原因：%s。'
                    '请如实告诉用户这一点，不要假装引用了他的书。'
                    % (via_ima.get('error') or '该知识库里没有匹配内容'))
                return local
            via_ima['hint'] = (via_ima.get('hint') or '') + '；内置示例语料里也没有相关内容'
            return via_ima

    # ③ 没配 ima：用内置语料
    return local_corpus_retrieve(query, subject)


def ima_enabled(creds=None):
    """该次请求是否会走 ima（供 UI 提示用）。"""
    c = ima_creds(creds)
    return bool(c['apiKey'] and c['clientId'])


def list_knowledge_bases(creds=None):
    """列出该凭证可用的 ima 知识库（名称 -> ID），供设置面板「按名称选择」。"""
    c = ima_creds(creds)
    if not c['apiKey'] or not c['clientId']:
        return {'ok': False, 'error': '未填写 ima API Key 或 Client ID'}
    j = None
    try:
        # 主接口：我可添加的知识库列表（limit <= 50）
        # 兜底：知识库搜索（注意 limit 必须在 (0,20]，否则返回 code 51）
        j = _ima_post('/get_addable_knowledge_base_list', {'limit': 50}, creds)
    except ima_index.ImaError:
        try:
            j = _ima_post('/search_knowledge_base', {'limit': 20, 'cursor': ''}, creds)
        except ima_index.ImaError as e:
            return {'ok': False, 'error': str(e)}
    try:
        d = j.get('data') or {}
        raw = (d.get('addable_knowledge_base_list') or d.get('info_list') or d.get('list')
               or d.get('knowledge_base_list') or d.get('items')
               or (d if isinstance(d, list) else []))
        items = []
        for it in raw:
            kb_id = (it.get('knowledge_base_id') or it.get('kb_id')
                     or it.get('id') or it.get('base_id') or '')
            name = it.get('name') or it.get('title') or it.get('knowledge_base_name') or '(未命名)'
            if kb_id:
                items.append({'id': kb_id, 'name': name})
        out = {'ok': len(items) > 0, 'items': items}
        if not items:
            out['raw'] = json.dumps(j, ensure_ascii=False)[:500]
        return out
    except Exception as e:
        return {'ok': False, 'error': str(e)}


def test_ima(creds=None):
    """测一次 ima：列出知识库 + 真正开始建索引（给 UI 的「测试连接」用）。

    判定标准：能列出知识库就算连通。如果绑了知识库，顺带报告「已解析几本 / 共几个文件」，
    以及哪些文件读不了、为什么 —— 这些信息对排查「抓不到我的书」最关键。
    """
    if not ima_enabled(creds):
        return {'ok': False, 'error': '未填写 ima API Key 或 Client ID'}

    kbs = list_knowledge_bases(creds)
    if not kbs.get('ok'):
        return {'ok': False, 'error': kbs.get('error') or '拉不到知识库列表'}

    out = {'ok': True, 'kbs': len(kbs['items']),
           'names': [k['name'] for k in kbs['items']][:12]}

    c = ima_creds(creds)
    creds = creds or {}
    kb_map = parse_kb_map(creds.get('imaKbMap') or os.environ.get('IMA_KB_MAP') or '')
    kb = list(kb_map.values())[0] if kb_map else ''
    if not kb:
        out['note'] = '连接正常，列出 %d 个知识库。还没给学科绑定知识库 —— 绑定后我才能读里面的书。' % len(kbs['items'])
        return out

    try:
        stats = ima_index.build_index(kb, c['apiKey'], c['clientId'], budget_seconds=25)
    except Exception as e:
        out['note'] = '连接正常，但建索引失败：%s' % e
        return out

    st = ima_index.index_status(kb)
    out.update({
        'index': st,
        'note': '连接正常。知识库共 %d 个文件，已解析 %d 本（%.1f 万字）%s'
                % (st['listed'], st['ready'], st['chars'] / 10000.0,
                   ('，还有 %d 个待解析（下次提问会继续）' % (st['listed'] - st['ready'] - st['unreadable'])
                    if st['listed'] - st['ready'] - st['unreadable'] > 0 else '')),
    })
    if st['unreadable']:
        out['note'] += '；有 %d 个读不了（多为扫描版/图片版/网页笔记），详见 index.unreadable_detail' % st['unreadable']
    return out
